src/Microsoft.ApiDesignGuidelines.Analyzers/Core/DoNotRaiseExceptionsInUnexpectedLocations.cs

```diff
@@ -1,59 +1,351 @@
+                // i.e public method inside an internal class is not considered public.
+                if (analyzeOnlyPublicMethods &&
+                    method.GetResultantVisibility() != SymbolVisibility.Public)
+                {
+                    return false;
+                }
+
+                return matchFunction(method, compilation);
+            }
+        }
+        
+        private static List<MethodCategory> GetMethodCategories(Compilation compilation)
+        {
+            var methodCategories = new List<MethodCategory>(12);
+            methodCategories.Add(new MethodCategory(IsPropertyGetter, true,
+                PropertyGetterRule,
+                WellKnownTypes.InvalidOperationException(compilation), WellKnownTypes.NotSupportedException(compilation)));
+
+            methodCategories.Add(new MethodCategory(IsIndexerGetter, true,
+                PropertyGetterRule,
+                WellKnownTypes.InvalidOperationException(compilation), WellKnownTypes.NotSupportedException(compilation),
+                WellKnownTypes.ArgumentException(compilation), WellKnownTypes.KeyNotFoundException(compilation)));
+
+            methodCategories.Add(new MethodCategory(IsEventAccessor, true,
+                HasAllowedExceptionsRule,
+                WellKnownTypes.InvalidOperationException(compilation), WellKnownTypes.NotSupportedException(compilation),
+                WellKnownTypes.ArgumentException(compilation)));
+
+            methodCategories.Add(new MethodCategory(IsGetHashCodeInterfaceImplementation, true,
+                HasAllowedExceptionsRule,
+                WellKnownTypes.ArgumentException(compilation)));
+
+            methodCategories.Add(new MethodCategory(IsEqualsOverrideOrInterfaceImplementation, true,
+                NoAllowedExceptionsRule));
+
+            methodCategories.Add(new MethodCategory(IsEqualityOperator, true,
+                NoAllowedExceptionsRule));
+
+            methodCategories.Add(new MethodCategory(IsGetHashCodeOverride, true,
+                NoAllowedExceptionsRule));
+
+            methodCategories.Add(new MethodCategory(IsToString, true,
+                NoAllowedExceptionsRule));
+
+            methodCategories.Add(new MethodCategory(IsImplicitCastOperator, true,
+                NoAllowedExceptionsRule));
+
+            methodCategories.Add(new MethodCategory(IsStaticConstructor, false,
+                NoAllowedExceptionsRule));
+
+            methodCategories.Add(new MethodCategory(IsFinalizer, false,
+                NoAllowedExceptionsRule));
+
+            methodCategories.Add(new MethodCategory(IMethodSymbolExtensions.IsDisposeImplementation, true,
+                NoAllowedExceptionsRule));
+
+            return methodCategories;
+        }
+
+        private static bool IsPropertyGetter(IMethodSymbol method, Compilation compilation)
+        {
+            return method.MethodKind == MethodKind.PropertyGet &&
```

Comment:
these seems general helper methods, can this be moved to extensions? like methodsymbolextensions?